controllers/hyperparametertuningjob/hyperparametertuningjob_controller_test.go

```diff
@@ -18,1680 +18,538 @@ package hyperparametertuningjob
-
-		// Run test
-		controller.Reconcile(request)
-
-		// Verify that the hpoTrainingJobSpawner was called
-		Expect(spawnMissingTrainingJobHpoJobs.Len()).To(Equal(1))
+	Context("HyperParameterTuningJob exists", func() {
 
-		// Verify that the hpoTrainingJobSpawner received the correct parameters.
-		actualHpoJob := spawnMissingTrainingJobHpoJobs.Front().Value.(hpojobv1.HyperparameterTuningJob)
-		Expect(*actualHpoJob.Spec.HyperParameterTuningJobName).To(Equal(*job.Spec.HyperParameterTuningJobName))
-		Expect(*actualHpoJob.Spec.Region).To(Equal(*job.Spec.Region))
-		Expect(actualHpoJob.ObjectMeta.GetNamespace()).To(Equal(job.ObjectMeta.GetNamespace()))
-
-		// Verify that the SageMaker request was made
-		Expect(receivedRequests.Len()).To(Equal(1))
-	})
-
-	Context("when the spec is updated", func() {
+		var expectedStatus sagemaker.HyperParameterTuningJobStatus
 
 		BeforeEach(func() {
-
-			// Get job so we can modify it.
-			err = k8sClient.Get(context.Background(), types.NamespacedName{
-				Namespace: job.ObjectMeta.Namespace,
-				Name:      job.ObjectMeta.Name,
-			}, job)
-			Expect(err).ToNot(HaveOccurred())
-
-			job.Spec.HyperParameterTuningJobConfig.ResourceLimits.MaxNumberOfTrainingJobs = ToInt64Ptr(*job.Spec.HyperParameterTuningJobConfig.ResourceLimits.MaxNumberOfTrainingJobs + 5)
-			err = k8sClient.Update(context.Background(), job)
-			Expect(err).ToNot(HaveOccurred())
-
+			shouldHaveFinalizer = true
 		})
 
-		It("should set status to failed", func() {
-			// Not failed
-			description.HyperParameterTuningJobStatus = sagemaker.HyperParameterTuningJobStatusInProgress
-
-			// Setup mock responses.
-			sageMakerClient := builder.
-				WithRequestList(&receivedRequests).
-				AddDescribeHyperParameterTuningJobResponse(description).
-				Build()
-
-			// Instantiate controller and reconciliation request.
-			controller := createHpoReconcilerForSageMakerClient(k8sClient, sageMakerClient, "1s")
-			request := CreateReconciliationRequest(job.ObjectMeta.Name, job.ObjectMeta.Namespace)
-
-			// Run test and verify expectations.
-			controller.Reconcile(request)
-
-			// Verify status is failed.
-			err = k8sClient.Get(context.Background(), types.NamespacedName{
-				Namespace: job.ObjectMeta.Namespace,
-				Name:      job.ObjectMeta.Name,
-			}, job)
-			Expect(job.Status.HyperParameterTuningJobStatus).To(Equal(string(sagemaker.HyperParameterTuningJobStatusFailed)))
+		Context("HyperParameterTuningJob has status 'InProgress'", func() {
+			BeforeEach(func() {
+				expectedStatus = sagemaker.HyperParameterTuningJobStatusInProgress
+				mockSageMakerClientBuilder.
+					AddDescribeHyperParameterTuningJobResponse(CreateDescribeOutputWithOnlyStatus(expectedStatus))
+			})
+
+			When("!HasDeletionTimestamp", func() {
+				It("Requeues after interval", func() {
+					ExpectRequeueAfterInterval(reconcileResult, reconcileError, pollDuration)
+				})
+
+				It("Updates status", func() {
+					ExpectStatusToBe(tuningJob, string(expectedStatus))
```

Comment:
Man, generics would be so powerful here. I wonder if there's a way we can rig our own generics by iterating over a set of preconditions that should match some assumption. It would look more like Go table-tests but would likely reduce how many tests we duplicate.